File: 39_Combination_Sum.py
from __future__ import print_function
from typing import List

# 39. Combination Sum
# https://leetcode.com/problems/combination-sum/
# Given an array of distinct integers candidates and a target integer target, return a list of all unique combinations of candidates where the chosen numbers sum to target.
# You may return the combinations in any order.

class Solution(object):
    def combinationSum(self, candidates, target):
        result = []

        def search(path, total):
            if total == target:
                result.append(path)
                return
            if total > target:
                return

            for i in range(len(candidates)):
                # Allow repeated use by passing same i
                new_path = path + [candidates[i]]
                search(new_path, total + candidates[i])

        # Sort to help with optimization (optional here)
        candidates.sort()
        search([], 0)
        return [combo for combo in result if sorted(combo) not in [sorted(r) for r in result[:result.index(combo)]]]
    
if __name__ == "__main__":
    s = Solution()
    candidates = [ 2, 3, 6, 7 ]
    target = 7
    result = s.combinationSum(candidates, target)
    print("result = ", result)  # [[2, 2, 3], [7]]
    
    candidates = [ 2, 3, 5 ]
    target = 8
    result = s.combinationSum(candidates, target)
    print("result = ", result)  # [[2, 2, 2, 2], [2, 3, 3], [3, 5]]

File: 39_Combination_Sum.cpp
#include <vector>
#include <iostream>
#include <functional>
#include <numeric>
using namespace std;

class Solution {
public:

    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        std::function<void(int, vector<int>&, vector<int>&, int, vector<vector<int>>&)> getSum;
        vector<vector<int>> sum{};
        vector<int> u_v{};

        std::sort(candidates.begin(), candidates.end());
        candidates.erase(unique(candidates.begin(), candidates.end()), candidates.end());

        getSum = [&](int start, vector<int>& candidates, vector<int>& u_v, int target, vector<vector<int>>& sum) -> void {
            if (target == 0) { 
                sum.push_back(u_v);
                return;
            }

            for (auto idx = start; idx != candidates.size(); ++idx) {
                int currVal = candidates[idx];
                if (currVal > target) {
                    break;
                }

                u_v.push_back(currVal);

                getSum(idx, candidates, u_v, target - currVal, sum);

                u_v.pop_back();
            }
        };

        getSum(0, candidates, u_v, target, sum);
        
        return sum;
    }
};

void print_result(vector<vector<int>>& vec) {
    for (auto _ : vec) {
        for (auto __ : _) {
            std::cout << __ << " ";
        }
        putchar('\n');
    }
    putchar('\n');
}

int main() {
    Solution soln = Solution();

    vector<int> candidates = {2,3,6,7};
    int target = 7;

    auto res = soln.combinationSum(candidates, target);
    print_result(res);
}
